test/Tests/src/Semantic/Conversion_Spec.enso

```diff
@@ -144,6 +152,57 @@ spec =
             Hello.formulate_with_to [ Hello.Say "Proper", Hello.Say "Type" ] . should_equal "ProperType"
             Hello.formulate_with_to [ Foo.Value "Perform", Bar.Value "Conversion" ] . should_equal "PERFORM conversion!"
 
+        Test.specify "Requesting Text & Foo" <|
+            check a (n : Text & Foo) = case a of
+                0 -> n.foo
+                1 -> n.take (First 3)
+
+            check 0 "Ahoj" . should_equal 4
+            check 1 "Ahoj" . should_equal "Aho"
+
+            # Boolean can be converted to Foo, but is not Text
+            fail = Panic.recover Type_Error <| check 0 True
+            fail . should_fail_with Type_Error
+
+        Test.specify "Requesting Foo & Not_Foo & Boolean" <|
+            check a (n : Foo & Not_Foo & Boolean) = case a of
+                0 -> n.foo
+                1 -> n.not
+                2 -> n.notfoo
+
+            check 0 True . should_be_true
+            check 1 True . should_be_false
+            check 2 True . should_be_true
+
+            fail = Panic.recover Type_Error <| check 0 "not a boolean"
+            fail . should_fail_with Type_Error
```

Comment:
Aren't we afraid that this kind of method resolution will introduce confusion?

If both `Foo` and `Not_Foo` have methods with the same name, which one is resolved will depend on the ordering of types there. IMO that will be confusing and not that easy to track down.

I'd rather detect ambiguity and require to narrow down to a single type, so that the reference is unambiguous.

I.e.
```
type A
    Value x
    foo self = self.x+100
    only_a self = self.x+1000
type B
    Value x
    foo self = self.x+200
    only_b self = self.x+2000

foo (v : A & B) =
    IO.println v.only_a # OK
    IO.println v.only_b # OK
    IO.println v.foo # IMO this should be an Ambiguous_Method_Resolution.Error saying that implementations from both A and B are available
    IO.println (v.to A . foo) # now it's unambiguous again
```

What do you think?

I'm not sure if what I'm proposing is better, but the current approach definitely introduces possible confusion. This is a big semantics change and I'm not aware of any discussions about it.

Reply:
> This is a big semantics change and I'm not aware of any discussions about it.

This PR only adds a runtime type checks for methods with intersection or union types. It does not define what intersection or union types are. Therefore, I'd say that there is in fact no semantic change at all. Moreover, this behavior is opt-in by specifying a particular method ascription.

Your observation with builtin nodes failing with `EnsoMultiValue` input is interesting, though. And worth further investigation.

Reply:
> This PR only adds a runtime type checks for methods with intersection or union types. It does not define what intersection or union types are. Therefore, I'd say that there is in fact no semantic change at all. Moreover, this behavior is opt-in by specifying a particular method ascription.

It introduces a new kind of value `EnsoMultiValue` which introduces a **new style of method resolution** (first type from types of the intersection type that contains a given method is chosen).

Yes, this is an opt-in change, that is true.

But it is available now. And once users start using it, it will be hard to change the semantics. And I feel like we did not have a chance to discuss the implications of e.g. the new method resolution style.

Reply:
> this should be an Ambiguous_Method_Resolution.Error saying that implementations from both A and B are available

Such an approach could lead to [CharSequence.isEmpty() situation](http://wiki.apidesign.org/wiki/BinaryIncompatibleDefaultMethods) - e.g. a situation where adding something into a type could break its usage. We haven't yet tried to evolve Enso libraries in a [99% compatible way](http://wiki.apidesign.org/wiki/BackwardCompatibility#Strictness), so it is hard to predict what is going to be optimal for that - in general I am [not fan of throwing errors in production](http://wiki.apidesign.org/wiki/Image:RinardOOPSLA06.pdf) when a viable, deterministic resolution exists. However yes, the check for ambiguities can be implemented with no slowdown with some work.

Btw. [even the sample test](https://github.com/enso-org/enso/pull/7769/files#diff-0ee393cab082d78e4b3dbe155decce832501c8cb080b9bbfb6961940dbfc7a08R613) contains such ambiguities as both types define `dict`. Based on such experience I [started work on `(x : Plus)` syntax](https://github.com/enso-org/enso/issues/6848#issuecomment-1711541161) that should make the _exact resolution_ natural for programmers.

Reply:
> > This PR only adds a runtime type checks for methods with intersection or union types. 
> > It does not define what intersection or union types are. 

A program that did run before and for some reason used `&` types, may no longer finish, but yield an error. E.g. yes, this changes semantics, but that's what we want: _either support `&` checks or remove `&` altogether_ just like `\`.

> Yes, this is an opt-in change, that is true.

There was no code using `&` yet - e.g. yes, it is an opt-in. I suggest to wait with widespread usage until [(x : Text)](https://github.com/enso-org/enso/issues/6848) destiny is set as the `x : Plus` syntax would make the usage far more easy.

> It introduces a new kind of value `EnsoMultiValue` 

Many engine PRs introduce new kind of value (for example #4023) and then we deal with consequences (like in #7348). There is nothing special on that and honestly, there is probably no other way to support `&` types in a dynamic language like Enso.

> which introduces a **new style of method resolution**:

Right. One has to resolve the methods somehow. Either:
- first type containing a given method is chosen
- last type ...
- random type ...
- error is raised when there are two types containing a method
- no method can be called on such multi value

Enough reviewers of this PR approved the _"first type"_ approach. However, as there are no usages yet, there is time for change.